### ISSUE TITLE
ERROR: Kernel configuration is invalid

### ISSUE THREAD

**Issue by davelister69** [seemoo-lab/bcm-public#3]
Hey guys,

I am attempting to patch the firmware on my nexus 5 with the nexmon patch provided.

I am running Xubuntu 14.0.4 and have NDK_ROOT set as described (I have android-ndk-r11c and the sdk correctly set in PATH).

I cloned the git repository and ran the following from the root directory:

source setup_env.sh
make reloadfirmware FWPATCH=nexmon

I initially received a gawk error (It's not included from an Xubuntu stock install, so I installed it), then re-run the last command.

I receive the following error:

```
make[2]: Entering directory `/home/anon/Downloads/bcm-public/kernel'

  ERROR: Kernel configuration is invalid.
         include/generated/autoconf.h or include/config/auto.conf are missing.
         Run 'make oldconfig && make prepare' on kernel src to fix it.
```

I ran 'make oldconfig && make prepare' in the kernel directory then re-run the reloadfirmware command from the root directory.

I receive the following error:

```
Make[2]: Entering directory `/home/anon/Downloads/bcm-public/kernel'

  WARNING: Symbol version dump /home/anon/Downloads/bcm-public/kernel/Module.symvers
           is missing; modules will have no dependencies and modversions.

  CC [M]  /home/anon/Downloads/bcm-public/firmware_patching/nexmon/bcmdhd/aiutils.o
  CC [M]  /home/anon/Downloads/bcm-public/firmware_patching/nexmon/bcmdhd/bcmsdh_sdmmc_linux.o
/bin/sh: 1: /home/anon/Downloads/bcm-public/kernel/scripts/recordmcount: not found
make[3]: *** [/home/anon/Downloads/bcm-public/firmware_patching/nexmon/bcmdhd/bcmsdh_sdmmc_linux.o] Error 127
make[3]: *** Waiting for unfinished jobs....
/bin/sh: 1: /home/anon/Downloads/bcm-public/kernel/scripts/recordmcount: not found
make[3]: *** [/home/anon/Downloads/bcm-public/firmware_patching/nexmon/bcmdhd/aiutils.o] Error 127
make[2]: *** [_module_/home/anon/Downloads/bcm-public/firmware_patching/nexmon/bcmdhd] Error 2
make[2]: Leaving directory `/home/anon/Downloads/bcm-public/kernel'
make[1]: *** [bcmdhd/bcmdhd.ko] Error 2
make[1]: Leaving directory `/home/anon/Downloads/bcm-public/firmware_patching/nexmon'
make: *** [nexmon] Error 2
```

I am a little confused. I assume I am not supposed to re-cross-compile the kernel provided as the target is fixed (the nexus 5).

Sorry if this is a simple question. Can you see what I am doing wrong?

**Comment by matthiasseemoo** [comment]
If you only intend to build the patched firmware without the kernel and boot image, you can simply run make in the "firmware_patching/nexmon" directory and then run "make install-firmware" to install the firmware.

According to your errors, it seems like sourcing the setup_env.sh did not work properly. We already deliver a working kernel config that should cross compile the kernel for your Nexus 5. However, if some environment variables defined in setup_env.sh are not set properly before making the kernel, it will try to compile it for your host platform on which you run the make command. This will result in errors about the kernel config. The best option would be to start fresh with the files of the repo and try again to follow the instructions.

**Comment by davelister69** [comment]
Hi. Thanks for your assistance.

I did as suggested and downloaded a fresh copy. I sourced the environment variables and ran make from the kernel folder. The kernel compiled successfully. I suspect I was running sudo which was dropping the environment variables. The -E switch preserved them.

If I go into the firmware_patching/nexmon directory and run make I receive the following error:

`anon@anon-desktop:~/Downloads/bcm-public/firmware_patching/nexmon$ make
make[1]: Entering directory `/home/anon/Downloads/bcm-public/firmware_patching/wrapper'
cat: /home/anon/Downloads/bcm-public/firmware_patching/nexmon/nexmon.generated.pre: No such file or directory
/bin/sh: 1: ../../buildtools/gcc-arm-none-eabi-5_4-2016q2/bin/arm-none-eabi-gcc: not found
/bin/sh: 1: ../../buildtools/gcc-arm-none-eabi-5_4-2016q2/bin/arm-none-eabi-gcc: not found
/bin/sh: 1: ../../buildtools/gcc-arm-none-eabi-5_4-2016q2/bin/arm-none-eabi-gcc: not found
make[1]: *** [/home/anon/Downloads/bcm-public/firmware_patching/nexmon/wrapper.o] Error 127
make[1]: Leaving directory `/home/anon/Downloads/bcm-public/firmware_patching/wrapper'
make: *** [wrapper.o] Error 2`

Weirdly enough the gcc file appears to be in the location bash is complaining about. Any suggestions?

**Comment by matthiasseemoo** [comment]
You should not need to compile it as root.

You can remove the @ after Q= in the Makefile to get more output. It could be a problem with the gcc plugin. Maybe you need to recompile it.

**Comment by DanielAW** [comment]
@davelister69 There should be a file called `buildtools/gcc-nexmon-plugin/nexmon.so`. You should also be able to build this by ourself just by executing `make` in the  `buildtools/gcc-nexmon-plugin` directory. 
You can also try a `make clean` in the `firmware_patching/nexmon/` folder before running `make` again.

**Comment by davelister69** [comment]
Hey. I managed to compile the kernel and the firmware patch. The issue was a few missing packages from a stock install of Xubuntu. I ran 'make install-firmware' and it ran through with no errors, eventually taking my wlan0 interface down.
I went into terminal on the phone and tried bringing it back up again but kept receiving:

`SIOCSIFFLAGS: Operation not permitted` 

I have only ever seen this when forgetting sudo, but I was running as root. I used adb to bring the interface back up:

`adb shell 'su -c "ifconfig wlan0 up"'`

which seemed to clear the problem. (I can now take the interface up or down)

However, I am attempting to put the interface into monitor mode using arimon-ng and it is failing:

`Airmon-ng start wlan0
ERROR adding monitor mode interface: command failed: Operation not supported (-95)`

It did mention wpa_supplicant when I first tried this, so I killed the process but it didn't resolve the issue.

I did read somewhere that wpa-supplicant interrupted the driver, and therefore that's why the full boot image was required.

I am basically looking to enable monitor/injection on a nethunter install. What is the path of least resistance to achieve this? Is it possible just to patch the firmware, or is it imperative to create the boot image?

Once again thanks for your assistance.

**Comment by DanielAW** [comment]
> However, I am attempting to put the interface into monitor mode using arimon-ng and it is failing

There is no need to manually set the interface in monitor mode using airomon-ng, our driver already sets the interface to monitor mode. The interface type of `wlan0` should be `link/ieee802.11/radiotap` if you ran `adb shell 'su -c "ip addr"'`. Using tcpdump (as mentioned in the README) for example should give you a reasonable output (no misinterpreted packets).

> Is it possible just to patch the firmware

Our driver modifications on the N5 basically just put the interface in monitor mode to make the aircrack suite work as intended. One could also use our [libfakeioctl library](https://github.com/seemoo-lab/bcm-public/tree/master/bootimg_src/src/libfakeioctl) in conjunction with LD_PRELOAD to make your tool (e.g. tcpdump) think the interface has the correct type.

**Comment by davelister69** [comment]
Running make install-firmware from the nexmon folder results in the driver being built, being pushed to the device and the wlan0 interface being taken down, like so:

```
make[1]: Leaving directory `/home/anon/Downloads/bcm-public/firmware_patching/nexmon'
adb shell 'su -c "mount -o rw,remount /system"' && \
	adb push fw_bcmdhd.bin /sdcard/ && \
	adb shell 'su -c "cp /sdcard/fw_bcmdhd.bin /vendor/firmware/fw_bcmdhd.bin"'
[100%] /sdcard/fw_bcmdhd.bin
adb shell 'su -c "ifconfig wlan0 down"'
```
Toasts appear on screen showing su privileges were granted for adb shell and the commands as listed above.  If I run `adb shell 'su -c "ip addr"'` as suggested I receive:

```
21: wlan0: <NO-CARRIER,BROADCAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state DORMANT qlen 1000
    link/ether 11:22:33:44:55:66 brd ff:ff:ff:ff:ff:ff
```
Does this mean the driver is incorrectly patched?

If I forgo airmon-ng and try running airodump-ng from the aircrack suite I receive:

```
ioctl(SIOCSIWMODE) failed: Operation not supported
Error setting monitor mode on wlan0
```
Assuming this is because the mode is being read incorrectly as suggested, I compiled the libfakeioctl library and pushed this to the sdcard.

If I run `LD_PRELOAD=libfakeioctl.so NEXMON_SA_FAMILY=ARPHRD_IEEE80211 airodump-ng wlan0` I receive:

`ERROR: ld.so: object 'libfakeioctl.so' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored`

I then receive the same error from airodump-ng as above. I ave only received library errors this way if the library is rejected at the OS level. I set the environment variables correctly prior to compilation.

Should I be able to patch the driver and successfully launch tools from the aircrack suite from the terminal on the phone?

Cheers
